spec/models/photosynthesis_measurement_spec.rb

```diff
@@ -1,5 +1,196 @@
+    @pm.internal_co2 = 1001.1001
+    @pm.vpd = 3.21
+    @pm.area_corr = 1
+    @pm.air_temp = 1.2
+    @pm.leaf_temp = 3.4
+    @pm.co2_reference = 1002.1002
+    @pm.co2_sample = 1003.1003
+    @pm.water_reference = 7.8
+    @pm.water_sample = 9.1
+    @pm.rh_reference = 91.91
+    @pm.rh_sample = 82.82
+    @pm.par_in = 1004.1004
+    @pm.pressure = 7.6
+  end
+
+  it 'is not valid on its own' do
+    expect(PhotosynthesisMeasurement.new).to_not be_valid
+  end
+
+  it 'can be valid' do
+    expect(@pm).to be_valid
+  end
+  
+  it 'is valid with type=AMAX' do
+    @pm.pm_type = 'AMAX'
+    expect(@pm).to be_valid
+  end
+
+  @badData = [
+    '',
+    'MILK',
+    'book',
+    'internal_co2',
+    'grock!',
+    '2',
+    973
+  ]
+  @badData.each do |val|
+    it "is invalid with a strange type" do
+      @pm.pm_type = val
+      expect(@pm).to_not be_valid
+    end
+  end
+  
+  it 'is invalid without a leaf part' do
+    @pm.leaf_part = nil
+    expect(@pm).to_not be_valid
+  end
+
+  it 'is invalid without a date' do
+    @pm.date = nil
+    expect(@pm).to_not be_valid
+  end
+  
+  it 'is invalid with a poorly formatted date' do
+    @pm.date = 'March 47th 2807'
+    expect(@pm).to_not be_valid
+  end
+
+  # time can't be validated in the model!
+  # if a badly formed time is put in the field it'll end up as nil
+  # which would be fine BUT time is sometimes nil anyway
+  # this will have to be validated in the importer
+
+  #it 'is invalid with a poorly formatted time' do
+  #  @pm.time = "12 O'clock"
+  #  expect(@pm).to_not be_valid
+  #end
+  
+  #it 'is valid with no time' do
+  #  @pm.time = nil
+  #  expect(@pm).to be_valid
+  #end
+  
+  @cols = [
+    'area_corr',
+    'photosynthesis_std',
+    'conductance_std',
+    'internal_co2',
+    'vpd',
+    'air_temp',
+    'leaf_temp',
+    'co2_reference',
+    'co2_sample',
+    'water_reference',
+    'water_sample',
+    'rh_reference',
+    'rh_sample',
+    'par_in',
+    'pressure'
+  ]
+  @v = -273.15
+  while @v < 0  do
+    @cols.each do |column|
+      it "is invalid with a negative #{column}" do
+        @pm[column] = @v
+        expect(@pm).to_not be_valid
+      end
+    end
+    @v = @v + 55
+  end
+
+  @cols = [
+    'photosynthesis_std',
+    'conductance_std',
+    'internal_co2',
+    'vpd',
+    'air_temp',
+    'leaf_temp',
+    'co2_reference',
+    'co2_sample',
+    'water_reference',
+    'water_sample',
+    'pressure'
+  ]  
+  @cols.each do |column|
+    it "is invalid with a zero #{column}" do
+      @pm[column] = 0
+      expect(@pm).to_not be_valid
+    end
+  end
+
+  @cols = [
+    'internal_co2',
+    'co2_reference',
+    'co2_sample'
+  ]  
+  @cols.each do |column|
+    it "is invalid with #{column} = 2000" do
+      @pm[column] = 2000
+      expect(@pm).to_not be_valid
+    end
+  end
+
+  @cols = [
+    'internal_co2',
+    'co2_reference',
+    'co2_sample',
+    'par_in'
+  ]
+  @v = 3000
+  while @v > 2000  do
```

Comment:
I'd never done unit testing with Ruby or Rails before and I was... erm... "exploring".
And it's ended up being, shall we say, thorough.